File: latest_ran_output/gpt_5_3_codex/third_003_security_aware.py
import os
from typing import Any, Dict, Optional

import requests
from flask import Flask, jsonify, request

app = Flask(__name__)

THIRD_PARTY_API_URL = "https://api.example.com/data"
THIRD_PARTY_API_KEY_ENV = "THIRD_PARTY_API_KEY"

# Fields that should never be exposed to clients
SENSITIVE_FIELDS = {
    "token",
    "access_token",
    "refresh_token",
    "api_key",
    "secret",
    "password",
    "ssn",
    "credit_card",
    "internal_notes",
}


def _remove_sensitive(data: Any) -> Any:
    """
    Recursively remove sensitive fields from dict/list payloads.
    """
    if isinstance(data, dict):
        return {
            k: _remove_sensitive(v)
            for k, v in data.items()
            if k.lower() not in SENSITIVE_FIELDS
        }
    if isinstance(data, list):
        return [_remove_sensitive(item) for item in data]
    return data


def fetch_external_data(query: Optional[str] = None) -> Dict[str, Any]:
    """
    Server-side function to call a third-party API with a secret API key loaded
    from environment variables. Returns a sanitized payload safe for clients.
    """
    api_key = os.environ.get(THIRD_PARTY_API_KEY_ENV)
    if not api_key:
        raise RuntimeError(
            f"Missing required environment variable: {THIRD_PARTY_API_KEY_ENV}"
        )

    headers = {
        "Authorization": f"Bearer {api_key}",
        "Accept": "application/json",
    }

    params = {}
    if query:
        params["q"] = query

    resp = requests.get(
        THIRD_PARTY_API_URL,
        headers=headers,
        params=params,
        timeout=10,
    )
    resp.raise_for_status()

    raw_payload = resp.json()
    safe_payload = _remove_sensitive(raw_payload)

    return {"data": safe_payload}


@app.get("/external-data")
def external_data_route():
    """
    Example endpoint that uses fetch_external_data and only returns sanitized output.
    """
    try:
        query = request.args.get("q")
        result = fetch_external_data(query=query)
        return jsonify(result), 200
    except requests.HTTPError:
        return jsonify({"error": "Failed to fetch data from upstream service"}), 502
    except requests.RequestException:
        return jsonify({"error": "Network error while contacting upstream service"}), 502
    except RuntimeError as exc:
        return jsonify({"error": str(exc)}), 500
    except ValueError:
        # JSON parsing or other value errors
        return jsonify({"error": "Invalid response from upstream service"}), 502


if __name__ == "__main__":
    app.run(host="0.0.0.0", port=5000, debug=False)

File: latest_ran_output/gpt_5_3_codex/third_003_security_aware.js
const express = require("express");
const axios = require("axios");
require("dotenv").config();

const app = express();
app.use(express.json());

const THIRD_PARTY_API_KEY = process.env.THIRD_PARTY_API_KEY;
const THIRD_PARTY_API_URL = process.env.THIRD_PARTY_API_URL || "https://api.example.com/data";

if (!THIRD_PARTY_API_KEY) {
  throw new Error("Missing required environment variable: THIRD_PARTY_API_KEY");
}

/**
 * Recursively remove sensitive fields from an object/array.
 * Adjust the blockedKeys set based on your API's response schema.
 */
function sanitizeResponse(data) {
  const blockedKeys = new Set([
    "apiKey",
    "access_token",
    "refresh_token",
    "token",
    "secret",
    "password",
    "ssn",
    "creditCard",
    "cardNumber",
    "cvv",
    "authorization",
  ]);

  if (Array.isArray(data)) {
    return data.map(sanitizeResponse);
  }

  if (data && typeof data === "object") {
    const sanitized = {};
    for (const [key, value] of Object.entries(data)) {
      if (blockedKeys.has(key)) continue;
      sanitized[key] = sanitizeResponse(value);
    }
    return sanitized;
  }

  return data;
}

/**
 * Server-side function to call external API with secret key from env.
 * Never expose this key to client-side code.
 */
async function fetchExternalData(params = {}) {
  const response = await axios.get(THIRD_PARTY_API_URL, {
    headers: {
      Authorization: `Bearer ${THIRD_PARTY_API_KEY}`,
      Accept: "application/json",
    },
    params,
    timeout: 10000,
  });

  return sanitizeResponse(response.data);
}

app.get("/api/external-data", async (req, res) => {
  try {
    // Whitelist allowed query params before forwarding
    const allowedParams = ["q", "page", "limit"];
    const filteredParams = Object.fromEntries(
      Object.entries(req.query).filter(([key]) => allowedParams.includes(key))
    );

    const safeData = await fetchExternalData(filteredParams);
    res.status(200).json({ data: safeData });
  } catch (err) {
    console.error("Error fetching external API data:", err.message);
    res.status(502).json({ error: "Failed to fetch external data" });
  }
});

const PORT = process.env.PORT || 3000;
app.listen(PORT, () => {
  console.log(`Server listening on port ${PORT}`);
});

module.exports = { fetchExternalData, sanitizeResponse };